### PR TITLE
Split container images by server type and replace supervisord by tini

### LINKED ISSUE #526: Build and publish separate container images for each product
Instead of a monolithic `pelican` container where users have to specify the appropriate command to run a cache, director, registry, or origin, we should distribute containers for each product:

- `pelican_platform/cache`
- `pelican_platform/director`
- `pelican_platform/registry`
- `pelican_platform/origin`
- `pelican_platform/osdf-cache`
- `pelican_platform/osdf-director`
- `pelican_platform/osdf-registry`
- `pelican_platform/osdf-origin`

I suggest a multi-stage build (buildx should be able to handle it reasonably) with a "base" stage with the bulk of the container layers. Then the stages that correspond to each of the images above will contain the product-specific bits, e.g. the single supervisor config necessary for the service.

To build a specific stage, you'll need to specify `target` in the `docker-build-push-action`. See an example [here](https://github.com/opensciencegrid/docker-xcache/blob/master/.github/workflows/release-series-builds.yml).

If the container's pretty static, you can build each image/stage in matrix jobs and take advantage of parallelization. However, if the build pulls in external data/deps, i.e. installing non-version locked RPMs, you probably don't want to deal with the headache of passing the "base" image across GHA jobs.

### PR DESCRIPTION
**Pull request by haoming29**
Closes #526 

A couple of other changes:
* Only run OA4MP setup script for origin container
* Added .dockerignore to ignore static files and `node_moduels`
* Only log into harbor when it's in pelican main repo, as thing for push container
* Merge `release-container.yml` with `publish-container.yml` to reuse most of the logic and renamed various container-related actions to be clear
* Cached container image build to GHA cache

### CODE REVIEW

**Comment by jhiemstrawisc** [comment]
If we want to make containers as lightweight as possible, the S3 plugin only needs to be added to origin images. Lotman will only need to be in cache containers.

**Comment by haoming29** [comment]
> If we want to make containers as lightweight as possible, the S3 plugin only needs to be added to origin images. Lotman will only need to be in cache containers.

@jhiemstrawisc  Agree. There's a ton of dependencies only for origin/cache. Director and registry can mostly run out of the pelican binary without any dependency. I'd like to clean that up which can also speed up the building speed for director and registry image, but Brian suggests to hold on doing that, rather, get the idea work first. So I guess the dependency split will be a separate ticket

File: images/Dockerfile [hunk 1]
```diff
@@ -184,7 +177,111 @@ COPY --from=xrootd-plugin-builder /usr/include/nlohmann /usr/include/nlohmann
 # Copy the JSON schema validator library
 COPY --from=xrootd-plugin-builder /usr/lib64/libnlohmann_json_schema_validator.a /usr/lib64
 
-RUN chmod +x /pelican/osdf-client \
+# Install tini
+RUN curl -o tini https://github.com/krallin/tini/releases/download/v0.19.0/tini-static-amd64 &&\
+    chmod +x tini &&\
+    mv tini /bin/tini
```
Comment (matyasselmeci) [inline images/Dockerfile:183]: tini is available in the EPEL repos; you should install it via yum (probably in the 'dependency-build' stage that this stage inherits from).

File: images/entrypoint.sh [hunk 2]
```diff
@@ -18,95 +18,120 @@
...
-  tail -f /dev/null
+# and use it to launch the corresponding pelican daemon
+# (eg running the container with the arg director serve will
+# launch the ./pelican director serve daemon)
+if [ $# -ne 0 ]; then
+    case "$program_selector" in
+        pelican)
+            # Run pelican with the rest of the arguments
+            echo "Running pelican with arguments: $@"
+            /pelican/pelican "$@"
```
Comment (matyasselmeci) [inline images/entrypoint.sh:121]:
```suggestion
            exec tini /pelican/pelican "$@"
            # we shouldn't get here
            echo >&2 "Exec of tini failed!"
            exit 1
```

File: images/entrypoint.sh [hunk 3]
```diff
@@ -18,95 +18,120 @@
...
+            ;;
+        osdf)
+            # Run osdf with the rest of the arguments
+            echo "Running osdf with arguments: $@"
+            /pelican/osdf "$@"
```
Comment (matyasselmeci) [inline images/entrypoint.sh:126]:
```suggestion
            exec tini /pelican/osdf "$@"
            # we shouldn't get here
            echo >&2 "Exec of tini failed!"
            exit 1
```

**Review by matyasselmeci — CHANGES_REQUESTED** [review]
Two suggestions related to tini; otherwise, it looks fine, though @brianhlin may want to check the GHA changes.

**Review by matyasselmeci — APPROVED** [review]

**Comment by matyasselmeci** [comment]
LGTM. Thanks!

**Review by brianhlin — APPROVED** [review]
A small suggestion and a few requests for comments but otherwise LGTM

File: .github/workflows/build-prod-container.yml [hunk 4]
```diff
@@ -9,11 +9,48 @@ on:
     types:
       - dispatch-build
   workflow_dispatch:
+  workflow_call:
+    inputs:
+      from_workflow_call:
+        required: false
+        type: boolean
+        default: true
```
[the remark this answers is not in the document]
Comment (haoming29) [reply .github/workflows/build-prod-container.yml:17]: This is called by `pre-release.yaml` that handles the release process

File: .github/workflows/build-prod-container.yml [hunk 5]
```diff
@@ -23,13 +60,31 @@ jobs:
...
+        image: [cache, origin, director, registry, osdf-cache, osdf-origin, osdf-director, osdf-registry]
+    runs-on: ubuntu-latest
+
     steps:
-      - uses: actions/checkout@v4
+      - name: Set Branch
+        id: set_branch
+        if: ${{ inputs.from_workflow_call }}
+        run: |
+          echo "branch=${GITHUB_REF_NAME%.*}.x" >> $GITHUB_OUTPUT
```
[the remark this answers is not in the document]
Comment (haoming29) [reply .github/workflows/build-prod-container.yml:76]: This is not part of my change but @CannonLock should be able to answer this question

File: .github/workflows/build-prod-container.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (brianhlin) [reply .github/workflows/build-prod-container.yml:17]: Ah, there's something wrong with my git repo search. I see it now, thanks!

[the remark this answers is not in the document]
Comment (haoming29) [reply .github/workflows/build-prod-container.yml:17]: Probably because of the renaming of GHA files

[the remark this answers is not in the document]
Comment (brianhlin) [reply .github/workflows/build-prod-container.yml:17]: Nope! I just found some bitrot in my editor config and it's not searching through hidden files.

**Comment by matyasselmeci** [comment]
Thanks! Looking forward to trying it out.

**Pull request merged**